build_product_images.py

```diff
@@ -125,48 +127,97 @@ def build_image_tags(image_name: str, image_version: str, product_version: str)
-    publish product images.
+    Generates a Bakefile (https://docs.docker.com/build/bake/file-definition/) describing how to build the whole image graph.
 
-    For local building, builder instances are supported.
+    build_and_publish_images() ensures that only the desired images are actually built.
+    """
+    targets = {}
+    groups = {}
+    contexts = {}
+    products = {product["name"]: product for product in conf.products}
+    product_names = list(products.keys())
+    for product_name, product in products.items():
+        product_targets = {}
+        for version_dict in product.get("versions"):
+            product_targets.update(bakefile_product_version_targets(args, product_name, version_dict, contexts, product_names))
+        groups[product_name] = {
+            "targets": list(product_targets.keys()),
+        }
+        targets.update(product_targets)
+    groups["default"] = {
+        "targets": list(groups.keys()),
+    }
+    return {
+        "target": targets,
+        "group": groups,
+    }
+
+
+def bakefile_target_name_for_product_version(product_name: str, version: str) -> str:
+    """
+    Creates a normalized Bakefile target name for a given (product, version) combination.
+    """
+    return f"{ product_name }-{ version.replace('.', '_') }"
+
+
+def bakefile_product_version_targets(args: Namespace, product_name: str, versions: Dict[str, str], contexts: Dict[str, str], product_names: List[str]):
+    """
+    Creates Bakefile targets defining how to build a given product version.
+
+    A product is assumed to depend on another if it defines a `versions` field with the same name as the other product.
     """
     image_name = f'{args.registry}/{args.organization}/{product_name}'
     tags = build_image_tags(
         image_name, args.image_version, versions["product"]
     )
     build_args = build_image_args(versions, args.image_version)
 
-    commands = [
-        "docker",
-        "buildx",
-        "build",
-        *build_args,
-        *tags,
-        "-f",
-        f"{ product_name }/Dockerfile",
-        "--platform",
-        ",".join(args.architecture),
-    ]
+    return {
+        bakefile_target_name_for_product_version(product_name, versions['product']): {
+            "dockerfile": f"{ product_name }/Dockerfile",
+            "tags": tags,
+            "args": build_args,
+            "platforms": args.architecture,
+            "context": ".",
+            "contexts": {f"stackable/image/{dep_name}": f"target:{bakefile_target_name_for_product_version(dep_name, dep_version)}" for dep_name, dep_version in versions.items() if dep_name in product_names},
+        },
+    }
 
-    if args.push:
-        commands.append(
-            "--push",
-        )
-    if not args.push and len(args.architecture) == 1:
-        commands.append(
-            "--load",
-        )
 
-    commands.append(".")
+def build_and_publish_image(args: Namespace, product_name: str, bakefile) -> List[List[str]]:
```

Comment:
```suggestion
def build_and_publish_image(args: Namespace, product_name: str, bakefile) -> List[Dict[str, Sequence[str]]]:
```

Reply:
I think I'd rather just remove the return type entirely then, the dict doesn't really have a meaningful "shape" that we can reasonably type like this (at least not without defining it as a class/namedtuple instead).

Reply:
1c5776f